walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,150 @@
+    
+    // sets up timeSlots array; anywhere left true has all the requestAttendees available
+    for (Event event : events) { 
+      mandatoryTimeSlots = convertToTimeSlots(requestAttendees, events);
+      allTimeSlots = convertToTimeSlots(allAttendees, events);    
+    }
+
+    ArrayList<Integer> possibleSlots = convertToSlots(mandatoryTimeSlots, numRequestSlots);
+    ArrayList<Integer> allSlots = convertToSlots(allTimeSlots, numRequestSlots);
+    
+    ArrayList<TimeRange> possibleTimes = convertToTimes(possibleSlots, numRequestSlots, numEvents);
+    ArrayList<TimeRange> allTimes = convertToTimes(allSlots, numRequestSlots, numEvents);
+
+    if (allTimes.isEmpty() && !possibleTimes.contains(TimeRange.WHOLE_DAY))
+        return possibleTimes;
+    return allTimes;
+  }
+  
+  public static boolean[] convertToTimeSlots (Collection<String> requestAttendees, Collection<Event> events) {
+    boolean[] timeSlots = new boolean[48];
+    Arrays.fill(timeSlots, Boolean.TRUE);
+    int numEvents = events.size();
+
+    for (Event event : events) {
+      Set<String> eventAttendees = event.getAttendees();
+      TimeRange when = event.getWhen();
+      int start = (when.start()) / 30;
+      int eventSlots = when.duration() / 30;
+
+      Iterator<String> eventAi = eventAttendees.iterator(); 
+      while(eventAi.hasNext()) {
+        numEvents--;
+        String eventAttendee = eventAi.next();
+
+        Iterator<String> requestAi = requestAttendees.iterator(); 
+        while (requestAi.hasNext()) {
+          String requestAttendee = requestAi.next();
+          if (eventAttendee.equals(requestAttendee)) {
+            for (int i = 0; i < eventSlots; i++) 
+              timeSlots[start + i] = false;
+            break;
+          }
+        }
+        if(numEvents == 0)
+            break;
+      }
+    }
+    return timeSlots;
+  }
+
+  // converts boolean availability to list of available slots
+  public static ArrayList<Integer> convertToSlots (boolean timeSlots[], int numRequestSlots) {
+    ArrayList<Integer> slots = new ArrayList<Integer>();
+    boolean add;
+    for (int i = 0; i < 48; i++) {
+      add = false;
+      if (timeSlots[i]) {
+        add = true;
+        // checks if the duration is longer than a single time slot
+        for (int j = 1; j < numRequestSlots; j++)
+          if ((i+j) < timeSlots.length && !timeSlots[i+j])
+            add = false;
+      }
+      if(add)
+        for (int j = 0; j < numRequestSlots; j++)
+          slots.add(i+j);
+    }
+    return slots;
+  }
+
+  // converts list of available slots to a list of TimeRanges 
+  public static ArrayList<TimeRange> convertToTimes (ArrayList<Integer> slots, int numRequestSlots, int numEvents) {
+    ArrayList<TimeRange> possibleTimes = new ArrayList<TimeRange>();
+    int start;
+    boolean done = false;
+
+    // if slots size is zero
+    if(slots.size() == 0) 
+      return possibleTimes;
+
+    start = slots.get(0)*30;
```

Comment:
nit: you can declare `start` here, no need to declare it above. That way you don't declare it before assigning it, which is less error prone.